scripts/constituents.py

```diff
@@ -525,25 +523,68 @@ def write_host_routines(cap, host, reg_funcname, init_funcname, num_const_funcna
+                cap.write(f"num_dyn_consts = num_dyn_consts + size(dyn_const_prop_{idx})", 3)
+                cap.write("end if", 2)
+            # end for
+            cap.write("num_consts = num_consts + num_dyn_consts", 2)
+            cap.comment("Pack dynamic_constituents array", 2)
+            cap.write(f"allocate(dynamic_constituents(num_dyn_consts), stat={herrcode})", 2)
+            cap.write(f"if ({herrcode} /= 0) then", 2)
+            cap.write(f"{herrmsg} = 'failed to allocate dynamic_constituents'", 3)
+            cap.write("else", 2)
+            cap.write("index_start = 1", 3)
+            for idx, scheme in enumerate(sorted(dyn_const_dict)):
+                cap.write(f"do index = 1, size(dyn_const_prop_{idx}, 1)", 3)
+                cap.write(f"dynamic_constituents(index + index_start - 1) = dyn_const_prop_{idx}(index)", 4)
+                cap.write("end do", 3)
+                cap.write(f"index_start = size(dyn_const_prop_{idx}, 1) + 1", 3)
+                cap.write(f"deallocate(dyn_const_prop_{idx})", 3)
+            # end for
+            cap.write("end if", 2)
+        # end if
+        cap.write(f"if ({herrcode} == 0) then", 2)
         cap.comment("Initialize constituent data and field object", 3)
-        stmt = "call {}%initialize_table(num_consts)"
-        cap.write(stmt.format(const_obj_name), 3)
+        stmt = f"call {const_obj_name}%initialize_table(num_consts)"
+        cap.write(stmt, 3)
         # Register host model constituents
         cap.comment("Add host model constituent metadata", 3)
         cap.write("do index = 1, size(host_constituents, 1)", 3)
         cap.write(f"if ({herrcode} == 0) then", 4)
         cap.write("const_prop => host_constituents(index)", 5)
-        stmt = "call {}%new_field(const_prop, {})"
-        cap.write(stmt.format(const_obj_name, obj_err_callstr), 5)
+        stmt = f"call {const_obj_name}%new_field(const_prop, {obj_err_callstr})"
+        cap.write(stmt, 5)
         cap.write("end if", 4)
         cap.write("nullify(const_prop)", 4)
-        cap.write("if ({} /= 0) then".format(herrcode), 4)
+        cap.write(f"if ({herrcode} /= 0) then", 4)
         cap.write("exit", 5)
         cap.write("end if", 4)
         cap.write("end do", 3)
         cap.write("end if", 2)
         cap.blank_line()
+        # Register dynamic constituents
+        if len(dyn_const_dict) > 0:
+           cap.comment("Add dynamic constituent properties", 2)
+           cap.write(f"if ({herrcode} == 0) then", 2)
+           cap.write(f"do index = 1, size(dynamic_constituents, 1)", 3)
+           cap.write(f"if ({herrcode} == 0) then", 4)
+           cap.write(f"const_prop => dynamic_constituents(index)", 5)
+           stmt = f"call {const_obj_name}%new_field(const_prop, {obj_err_callstr})"
+           cap.write(stmt, 5)
+           cap.write("end if", 4)
+           cap.write("nullify(const_prop)", 4)
+           cap.write(f"if ({herrcode} /= 0) then", 4)
+           cap.write("exit", 5)
```

Comment:
Should we just `exit` the loop here, or fully `return` from the subroutine?

Reply:
returns added across the board for the generated code.